src/coreclr/debug/ee/dactable.cpp

```diff
@@ -23,45 +23,133 @@
+#define DEFINE_DACVAR_NODUMP(size, id, var) static_assert(!is_type_template_instantiation<decltype(var), Volatile>::m_value, "DAC variables defined with DEFINE_DACVAR_NODUMP must not be instantiations of Volatile<T>.");
+#define DEFINE_DACVAR_VOLATILE(size, id, var) static_assert(is_type_template_instantiation<decltype(var), Volatile>::m_value, "DAC variables defined with DEFINE_DACVAR_VOLATILE must be instantiations of Volatile<T>.");
+#include "dacvars.h"
+#undef DEFINE_DACVAR_VOLATILE
+#undef DEFINE_DACVAR_NODUMP
+#undef DEFINE_DACVAR
 }
+#else
+// Only dynamically initialize on non-MSVC builds since we can't handle symbol aliasing
+// the same way we do with MSVC to statically initialize the DAC table.
+DLLEXPORT DacGlobals g_DacTable;
 
-// Initializes the non-SVR table entries
-void DacGlobals::InitializeEntries(TADDR baseAddress)
+void DacGlobals::Initialize()
 {
-#define DEFINE_DACVAR(id_type, size, id, var)                   id = PTR_TO_TADDR(&var) - baseAddress;
-#define DEFINE_DACVAR_NO_DUMP(id_type, size, id, var)           id = PTR_TO_TADDR(&var) - baseAddress;
+    
+#define DEFINE_DACVAR(size, id, var) static_assert(!is_type_template_instantiation<decltype(var), Volatile>::m_value, "DAC variables defined with DEFINE_DACVAR must not be instantiations of Volatile<T>.");
+#define DEFINE_DACVAR_NODUMP(size, id, var) static_assert(!is_type_template_instantiation<decltype(var), Volatile>::m_value, "DAC variables defined with DEFINE_DACVAR_NODUMP must not be instantiations of Volatile<T>.");
+#define DEFINE_DACVAR_VOLATILE(size, id, var) static_assert(is_type_template_instantiation<decltype(var), Volatile>::m_value, "DAC variables defined with DEFINE_DACVAR_VOLATILE must be instantiations of Volatile<T>.");
 #include "dacvars.h"
+#undef DEFINE_DACVAR_VOLATILE
+#undef DEFINE_DACVAR_NODUMP
+#undef DEFINE_DACVAR
 
+#define DEFINE_DACVAR(size, id, var)                   id = PTR_TO_TADDR(&var);
+#define DEFINE_DACVAR_VOLATILE(size, id, var)          id = PTR_TO_TADDR(&var.m_val);
+#define DEFINE_DACVAR_NO_DUMP(size, id, var)           id = PTR_TO_TADDR(&var);
+#undef DEFINE_DACVAR_VOLATILE
+#undef DEFINE_DACVAR_NODUMP
+#undef DEFINE_DACVAR
+#include "dacvars.h"
+#undef DEFINE_DACVAR
+#undef DEFINE_DACVAR_VOLATILE
+#undef DEFINE_DACVAR_NO_DUMP
+#define DEFINE_DACGFN(func) PTR_TO_TADDR(&func),
+#define DEFINE_DACGFN_STATIC(class, func) PTR_TO_TADDR(&class::func),
+#include "gfunc_list.h"
+#undef DEFINE_DACGFN
+#undef DEFINE_DACGFN_STATIC
 #define VPTR_CLASS(name) \
     { \
         void *pBuf = _alloca(sizeof(name)); \
         name *dummy = new (pBuf) name(0); \
-        name##__vtAddr = PTR_TO_TADDR(*((PVOID*)dummy)) - baseAddress; \
+        name##__vtAddr = PTR_TO_TADDR(*((PVOID*)dummy)); \
     }
 #define VPTR_MULTI_CLASS(name, keyBase) \
```

Comment:
Is `VPTR_MULTI_CLASS` used anywhere?

Reply:
It is not, and it looks like the last time it was used was with NGEN. I'll remove it.